README.md
# dbs-vue-cli

## Project setup
```
npm install
```

### Compiles and hot-reloads for development
```
npm run serve
```

### Compiles and minifies for production
```
npm run build
```

### Run your tests
```
npm run test
```

### Lints and fixes files
```
npm run lint
```

### Customize configuration
See [Configuration Reference](https://cli.vuejs.org/config/).


CHANGELOG.md
# Changelog - Digital Benefit Sheet
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [4.1.0] - 2021-05-06
### Changed
- $40 activation is now $35 activation

### Fixed
- Fixed the discounts on the 5GB, 10GB and $110 Unlimited plans

## [4.0.0] - 2021-03-30
### Added
- Logs
- Student discount
- DPP Full Retail
- Credits Total
- Visa Card

### Changed
- Cookies should now expire at midnight. This will prevent "sessions" from continuing to the next day.
- Rearranged the Mix and Match plans to display from highest cost to lowest cost
- Visually separated Mix and Match 3.0, 2.0, and 1.0
- Moved discounts to its own tab
- "Common discounts" such as Military, First Responder, Nurse, etc., now can be selected using radio buttons
	- This helps clarify that only one can be selected, and makes it easier to add/remove discounts going forward
- Added a new format for the DPP page
	- This new format takes up less space, so you can add more device payments
	- The new "popup" allows more space for the added "Total" option for both dpp and credits, as well as "Manual" mode
	- This should also be easier for someone new to learn

## [3.3.0] - 2021-02-18
### Added
- Added Business tablet plan updates
- Added LTE Home Internet

### Changed
- The Business DPP Section now matches the consumer side

### Fixed
- Password reset should now be fixed

## [3.2.0] - 2021-02-01
### Added
- Added the new Loyalty Mix and Match plans

### Fixed
- Removed the Gizmo discount from the Do More Unlimited plan

## [3.1.0] - 2020-11-20
### Added
- Business plans now show the labels of chosen plans and devices on the Benefit Sheet (the same way Consumer plans work)

### Fixed
- Fixed the color mismatch on the benefits when printing the Benefit Sheet

## [3.0.3] - 2020-11-19
### Fixed
- It turns out that many users were putting in their emails with capital letters. This was causing the site to reject the credentials. This has now been fixed by converting the entire email to lowercase before authenticating the user
- The military discount on Business accounts has giving twice the correct amount. This is now fixed

## [3.0.2] - 2020-11-18
### Fixed
- Many users were running into traffic-related issues
	- To fix this, instead of updating devices every time a user navigates to the DPP page, the site will now save the list of devices in the state
	- The benefit is that users will be ***far*** less likely to run into traffic-related issues
	- The downside is that **users will now need to logout and login in order to see updated device pricing**
	- Users are forced to log in at least once a day, so this shouldn't be much of an issue

## [3.0.1] - 2020-11-12
### Fixed
- Pull thru devices on Business plans weren't editing the correct values

## [3.0.0] - 2020-11-10
### Added
- Added devices to the database
	- There is now a dropdown list of devices to the DPP section
	- When you choose a device, it populates the DPP with the full retail divided by 24 months
	- You can still add DPP manually, either by leaving the device input blank, or choosing the "Manual" option
	- **Admins can edit devices from the admin panel. I am unable to stay on top of any changes by myself, so please make sure to update whenever changes are made to the Pricing and Ordering Sheet**

### Changed
- Changed servers
	- Benefits
		- Better scalability
		- Better security
		- More adaptable to future needs
		- More control
	- Please let me know if you notice any performance issues
- The admin panel is now an entirely separate app
	- This will help with performance, as well as help me maintain the project

## [2.8.0] - 2020-10-02
### Added
- Added the option to pull last month's usage csv in the admin panel

## [2.7.0] - 2020-09-17
### Added
- Added the new Unlimited Plus plan options for tablets and jetpacks
- Added a "Notes" section to the "accordion" style of the Benefit Sheet
- Added Disney+, Hulu, ESPN+ as a benefit in the dropdown list

## [2.6.0] - 2020-08-20
### Added
- Added Mix and Match 3.0 plans
- Added a CSV download to the admin panel

## [2.5.1] - 2020-07-21
### Changed
- There is some confusion about what discounts apply to the 5GB plan, so I have temporarily removed any discounts (other than autopay) from the calculations
	- I will add these back once we have clarification on what does/doesn't qualify

## [2.5.0] - 2020-07-13
### Added
- Added labeling above the options that specify what devices are a part of each option
	- This is limited to phone plans and pull thru

### Fixed
- Fixed a spelling error

## [2.4.0] - 2020-07-02
### Added
- Added the new 5GB and 10GB plans

## [2.3.0] - 2020-06-16
### Added
- Added some more benefits to the dropdown on the Benefit Sheet Page

### Fixed
- I am now rounding the displayed tax percent on the Benefit Sheet to the nearest third decimal point
	- This is because of a known issue when dealing with decimal points when programming
	- [Here](https://floating-point-gui.de/) is more information on this issue if you're curious

## [2.2.0] - 2020-06-09
### Added
- Added the Loyalty 55+ plan
	- Since Jetpacks are not allowed on this plan, Option 1 and Option 2 will dynamically remove Jetpacks as an option when this plan is selected
- Added the Loyalty Go Unlimited plan
- Added an "other" section for miscellaneous costs
- Added a "tax" option
	- You can find this under the dropdown menu on the "Benefit Sheet" page (the same place "Print", "Delete Benefit Sheet", and "Delete All" are located)

## [2.1.1] - 2020-04-17
### Fixed
- Price correction for HumX on Business Accounts
	- Please logout and log back in to see the changes

## [2.1.0] - 2020-03-11
### Added

src/benefitSheet/vue.js
new Vue({
  el: '#sidebar',
  data: {
    version: '0.2.2'
  },
  components: {
    'app-sidebar': AppSidebar
  }
});

new Vue({
  el: '#app',
  components: {
    'app-autocomplete': autocomplete
  },

  data: {

    isLoaded: false,

    mainObject: {

      activetab: 1,
      numberOfPhones: 0,
      chosenPlan: "Unlimited",
      tmp: "$0",
      tmpConvertedFromString: 0,
      military: false,
      responder: false,
      discount: "0%",
      autopay: false,

      benefits: ['TEST', '', '', '', '', ''],

      tabsArray: [
        { name: 'planTabActive', value: true },
        { name: 'pullthruTabActive', value: false },
        { name: 'dppTabActive', value: false }
      ],

      mixAndMatchPlansArray: [
        {
          name: 'Go',
          numberOfPhones: 0,
          1: { lines: 1, autopay: 75, noAutopay: 80 },
          2: { lines: 2, autopay: 65, noAutopay: 70 },
          3: { lines: 3, autopay: 50, noAutopay: 55 },
          4: { lines: 4, autopay: 40, noAutopay: 45 }
        },
        {
          name: 'Beyond',
          numberOfPhones: 0,
          1: { lines: 1, autopay: 85, noAutopay: 90 },
          2: { lines: 2, autopay: 80, noAutopay: 85 },
          3: { lines: 3, autopay: 60, noAutopay: 65 },
          4: { lines: 4, autopay: 50, noAutopay: 55 }
        },
        {
          name: 'Above',
          numberOfPhones: 0,
          1: { lines: 1, autopay: 95, noAutopay: 100 },
          2: { lines: 2, autopay: 90, noAutopay: 95 },
          3: { lines: 3, autopay: 70, noAutopay: 75 },
          4: { lines: 4, autopay: 60, noAutopay: 65 }
        }
      ],

      militaryDiscountAmount: {
        1: 15,
        2: 35,
        3: 40,
        4: 40
      },

      oldUnlimitedPlansArray: [
        { name: '$65 Unlimited', autopay: 60, noAutopay: 65 },
        { name: '$110 Unlimited', autopay: 100, noAutopay: 110 }
      ],

      tieredPlansArray: [
        { name: '2GB', autopay: 35, noAutopay: 35 },
        { name: '4GB', autopay: 50, noAutopay: 50 },
        { name: '5GB', autopay: 35, noAutopay: 40 },
        { name: '8GB', autopay: 70, noAutopay: 70 },
        { name: '12GB', autopay: 80, noAutopay: 80 },
        { name: '16GB', autopay: 90, noAutopay: 90 },
        { name: '24GB', autopay: 110, noAutopay: 110 }
      ],

      twoyear: {
        value: 0,
        cost: 20
      },

      basic: {
        value: 0,
        cost: 30
      },

      lineAccess: [
        { name: 'Tablet', tiered: 10, unlimited: 20, value: 0 },
        { name: 'Jetpack', tiered: 10, unlimited: 20, value: 0 },
        { name: 'Hum+', tiered: 10, unlimited: 10, value: 0 },
        { name: 'HumX', tiered: 15, unlimited: 20, value: 0 },
        { name: 'Homephone', tiered: 20, unlimited: 20, value: 0 },
        { name: 'Gizmo', tiered: 5, unlimited: 5, value: 0 },
        { name: 'Smartwatch', tiered: 10, unlimited: 10, value: 0 },
        { name: 'Arlo', tiered: 10, unlimited: 20, value: 0 }
      ],

      existingDPPValues: [
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" }
      ],

      existingCreditValues: [
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" },
        { value: "" }
      ]

    }

  },

  mounted() {
    const parsed = JSON.stringify(this.mainObject);
    localStorage.setItem('defaultMainObject', parsed);

    if (localStorage.getItem('mainObject')) {
      try {
        this.mainObject = JSON.parse(localStorage.getItem('mainObject'));
      } catch(e) {
        localStorage.removeItem('mainObject');
      }
    }

    this.isLoaded = true;
  },

  methods: {

    navbarClick: function(button) {
      var i;
      for (i = 0; i < mainObject.tabsArray.length; i++)  {
        mainObject.tabsArray[i].value = false;
        if (button === mainObject.tabsArray[i].name) {
          mainObject.tabsArray[i].value = true;
        } else {
          mainObject.tabsArray[i].value = false;
        }
      }
    },

    mixAndMatchPlanMath: function() {
      var localTotal = 0;
      var localNumberOfPhones = this.computedNumberOfPhones;

      if (localNumberOfPhones < 1) {
        return 0;
      } else if (localNumberOfPhones > 4) {
        localNumberOfPhones = 4;
      }

      for (i=0; i<this.mainObject.mixAndMatchPlansArray.length; i++) {
        if (this.mainObject.autopay === true) {
          localTotal = localTotal + (this.mainObject.mixAndMatchPlansArray[i].numberOfPhones * this.mainObject.mixAndMatchPlansArray[i][localNumberOfPhones].autopay);
        } else if (this.mainObject.autopay === false) {
          localTotal = localTotal + (this.mainObject.mixAndMatchPlansArray[i].numberOfPhones * this.mainObject.mixAndMatchPlansArray[i][localNumberOfPhones].noAutopay);
        }
      }

      if (this.mainObject.military === true || this.mainObject.responder === true) {
        localTotal = localTotal - this.mainObject.militaryDiscountAmount[localNumberOfPhones];
      }

      localTotal = localTotal + (this.mainObject.twoyear.value * this.mainObject.twoyear.cost); 

      localTotal = localTotal + (this.mainObject.basic.value * this.mainObject.basic.cost); 

      localNumberOfPhones = 0;

      return localTotal;
    },

    oldUnlimitedAndNewVerizonPlans: function() {
      if (this.computedNumberOfPhones <= 0) {
        return 0;
      }

      var localTotal = 0;

      for (i=0; i<this.mainObject.oldUnlimitedPlansArray.length; i++) {
        if(this.mainObject.chosenPlan === this.mainObject.oldUnlimitedPlansArray[i].name) {
          if (this.mainObject.autopay === true) {
            localTotal = localTotal + this.mainObject.oldUnlimitedPlansArray[i].autopay;
          } else if (this.mainObject.autopay === false) {
            localTotal = localTotal + this.mainObject.oldUnlimitedPlansArray[i].noAutopay;
          }

          if (this.mainObject.military === true || this.mainObject.responder === true) {
            localTotal = localTotal * .85;
          }
        }
      }

      for (i=0; i<this.mainObject.tieredPlansArray.length; i++) {
        if(this.mainObject.chosenPlan === this.mainObject.tieredPlansArray[i].name) {
          if (this.mainObject.autopay === true) {
            localTotal = localTotal + this.mainObject.tieredPlansArray[i].autopay;
          } else if (this.mainObject.autopay === false) {
            localTotal = localTotal + this.mainObject.tieredPlansArray[i].noAutopay;                       
          }

          if (this.mainObject.military === true || this.mainObject.responder === true) {
            localTotal = localTotal * .85;
          } else if (parseInt(this.mainObject.discount) > 0) {
            localTotal = localTotal * (.01 * (100 - parseInt(this.mainObject.discount)));
          }
        }
      }

      localTotal = localTotal + (this.mainObject.twoyear.value * this.mainObject.twoyear.cost); 

      localTotal = localTotal + (this.mainObject.basic.value * this.mainObject.basic.cost); 

      return localTotal;
    },

    saveToLocalStorage: function() {
      const parsed = JSON.stringify(this.mainObject);
      localStorage.setItem('mainObject', parsed);
    },

    restoreDefaultValues: function() {
      if (localStorage.getItem('defaultMainObject')) {
        try {
          this.mainObject = JSON.parse(localStorage.getItem('defaultMainObject'));
        } catch(e) {
          localStorage.removeItem('defaultMainObject');
        }
      }
    }

  },

  computed: {

    tieredPlansArrayReversed: function() {
      return this.mainObject.tieredPlansArray.slice().reverse();
    },

    oldUnlimitedPlansArrayReversed: function() {
      return this.mainObject.oldUnlimitedPlansArray.slice().reverse();
    },

    total: function() {
      if (this.isLoaded === false) {
        console.log('Loading');
        return 'Loading';
      } else if (this.isLoaded === true) {
        console.log('Loaded');
        this.saveToLocalStorage();

        if (this.computedNumberOfPhones <= 0) {
          return 'Add Phones';
        }

        return (this.planTotal + this.lineAccessTotal + this.tmpTotal + this.dppTotal).toFixed(2);
      }
    },

    textBeforeTotal: function() {
      if (this.total !== 'Add Phones') {
        return 'Total: $';
      }
    },

    benefitSheetTotalOption1: function() {
      if (this.total === 'Add Phones') {
        return this.total;
      } else if (this.total !== 'Add Phones') {
        return '$' + this.total + ' plus tax';
      }
    },

    planTotal: function() {
      if (this.mainObject.chosenPlan === 'Unlimited') {
        return this.mixAndMatchPlanMath();
      } else {
        return this.oldUnlimitedAndNewVerizonPlans();
      }
    },

    lineAccessTotal: function() {
      var localTotal = 0;

      if (this.mainObject.chosenPlan === 'Unlimited') {
        for (i=0; i<this.mainObject.lineAccess.length; i++) {
          localTotal = localTotal + (this.mainObject.lineAccess[i].unlimited * this.mainObject.lineAccess[i].value);
        }

        if (this.mainObject.military === true) {
          localTotal = localTotal * .85;
        }
      }

      for (i=0; i<this.mainObject.oldUnlimitedPlansArray.length; i++) {
        if (this.mainObject.chosenPlan === this.mainObject.oldUnlimitedPlansArray[i].name) {
          for (i=0; i<this.mainObject.lineAccess.length; i++) {
            localTotal = localTotal + (this.mainObject.lineAccess[i].unlimited * this.mainObject.lineAccess[i].value);
          }
          localTotal = localTotal + (this.computedNumberOfPhones * 20);
        }
      }

      for (i=0; i<this.mainObject.tieredPlansArray.length; i++) {
        if (this.mainObject.chosenPlan === this.mainObject.tieredPlansArray[i].name) {
          for (i=0; i<this.mainObject.lineAccess.length; i++) {
            localTotal = localTotal + (this.mainObject.lineAccess[i].tiered * this.mainObject.lineAccess[i].value);
          }
          localTotal = localTotal + (this.computedNumberOfPhones * 20);
        }
      }

      return localTotal;

    },

    tmpTotal: function() {
      var localTotal = 0;

      this.mainObject.tmpConvertedFromString = parseInt(this.mainObject.tmp.replace("$",""));
      localTotal = localTotal + this.mainObject.tmpConvertedFromString;

      return localTotal;
    },

    dppTotal: function() {
      var localTotal = 0;

      for (i=0; i<this.mainObject.existingDPPValues.length; i++) {
        if (isNaN(parseInt(this.mainObject.existingDPPValues[i].value))) {
        } else {
          localTotal = localTotal + this.mainObject.existingDPPValues[i].value;
        }
      }

      for (i=0; i<this.mainObject.existingCreditValues.length; i++) {
        if (isNaN(parseInt(this.mainObject.existingCreditValues[i].value))) {
        } else {
          localTotal = localTotal - this.mainObject.existingCreditValues[i].value;
        }
      }

      return localTotal;
    },

    computedNumberOfPhones: function() {
      var localTotal = 0;

      if (this.mainObject.chosenPlan === 'Unlimited') {

        for (i=0; i < this.mainObject.mixAndMatchPlansArray.length; i++) {
          localTotal = localTotal + this.mainObject.mixAndMatchPlansArray[i].numberOfPhones;
        }

        this.mainObject.numberOfPhones = 0;

      } else if (this.mainObject.chosenPlan !== 'Unlimited') {

        for (i=0; i < this.mainObject.mixAndMatchPlansArray.length; i++) {
          this.mainObject.mixAndMatchPlansArray[i].numberOfPhones = 0;
        }

        localTotal = this.mainObject.numberOfPhones;
      }

      if (localTotal == null) {
        localTotal = 0;
      }

      return localTotal;
    }

  }
});
